apps/pipelines/nodes/nodes.py

```diff
@@ -622,3 +626,118 @@ def _get_assistant_runnable(self, assistant: OpenAiAssistant, session: Experimen
+    code: str = Field(
+        default=DEFAULT_FUNCTION,
+        description="The code to run",
+        json_schema_extra=UiSchema(widget=Widgets.code),
+    )
+
+    @field_validator("code")
+    def validate_code(cls, value, info: FieldValidationInfo):
+        if not value:
+            value = DEFAULT_FUNCTION
+        try:
+            byte_code = compile_restricted(
+                value,
+                filename="<inline code>",
+                mode="exec",
+            )
+            custom_locals = {}
+            exec(byte_code, {}, custom_locals)
+
+            try:
+                main = custom_locals["main"]
+            except KeyError:
+                raise SyntaxError("You must define a 'main' function")
+
+            for name, item in custom_locals.items():
+                if name != "main" and inspect.isfunction(item):
+                    raise SyntaxError(
+                        "You can only define a single function, 'main' at the top level. "
+                        "You may use nested functions inside that function if required"
+                    )
+
+            if list(inspect.signature(main).parameters) != ["input", "kwargs"]:
+                raise SyntaxError("The main function should have the signature main(input, **kwargs) only.")
+
+        except SyntaxError as exc:
+            raise PydanticCustomError("invalid_code", "{error}", {"error": exc.msg})
+        return value
+
+    def _process(self, input: str, state: PipelineState, node_id: str) -> PipelineState:
+        function_name = "main"
+        byte_code = compile_restricted(
+            self.code,
+            filename="<inline code>",
+            mode="exec",
+        )
+
+        custom_locals = {}
+        custom_globals = self._get_custom_globals()
+        try:
+            exec(byte_code, custom_globals, custom_locals)
+            result = str(custom_locals[function_name](input))
+        except Exception as exc:
+            raise PipelineNodeRunError(exc) from exc
+        return PipelineState.from_node_output(node_id=node_id, output=result)
+
+    def _get_custom_globals(self):
+        from RestrictedPython.Eval import (
+            default_guarded_getitem,
+            default_guarded_getiter,
+        )
+
+        custom_globals = safe_globals.copy()
+        custom_globals.update(
+            {
+                "__builtins__": self._get_custom_builtins(),
+                "json": json,
+                "datetime": datetime,
+                "time": time,
+                "_getitem_": default_guarded_getitem,
+                "_getiter_": default_guarded_getiter,
+                "_write_": lambda x: x,
+            }
+        )
+        return custom_globals
+
+    def _get_custom_builtins(self):
+        allowed_modules = {
+            "json",
+            "re",
+            "datetime",
+            "time",
+        }
+        custom_builtins = safe_builtins.copy()
+        custom_builtins.update(
+            {
```

Comment:
(not for this PR) It would be nice to have some way of telling/showing the user which builtin modules and imports are available. Maybe we can utilize codemirror's [autocomplete](https://codemirror.net/examples/autocompletion/)